GiniVision/Classes/Core/Screens/Camera/CameraPreviewViewController.swift

```diff
@@ -0,0 +1,292 @@
+//
+//  CameraPreviewViewController.swift
+//  GiniVision
+//
+//  Created by Enrique del Pozo Gómez on 8/10/18.
+//
+
+import UIKit
+import AVFoundation
+
+protocol CameraPreviewViewControllerDelegate: class {
+    func cameraPreview(_ viewController: CameraPreviewViewController,
+                       didDetect qrCodeDocument: GiniQRCodeDocument)
+}
+
+final class CameraPreviewViewController: UIViewController {
+    
+    weak var delegate: CameraPreviewViewControllerDelegate?
+    
+    fileprivate let giniConfiguration: GiniConfiguration
+    fileprivate typealias FocusIndicator = UIImageView
+    fileprivate var camera: Camera?
+    fileprivate var cameraState = CameraState.notValid
+    fileprivate var defaultImageView: UIImageView?
+    fileprivate var focusIndicatorImageView: UIImageView?
+    fileprivate let interfaceOrientationsMapping: [UIInterfaceOrientation: AVCaptureVideoOrientation] = [
+        .portrait: .portrait,
+        .landscapeRight: .landscapeRight,
+        .landscapeLeft: .landscapeLeft,
+        .portraitUpsideDown: .portraitUpsideDown
+    ]
+    
+    fileprivate var cameraFocusSmall: UIImage? {
+        return UIImageNamedPreferred(named: "cameraFocusSmall")
+    }
+    
+    fileprivate var cameraFocusLarge: UIImage? {
+        return UIImageNamedPreferred(named: "cameraFocusLarge")
+    }
+    
+    fileprivate var defaultImage: UIImage? {
+        return UIImageNamedPreferred(named: "cameraDefaultDocumentImage")
+    }
+    
+    fileprivate enum CameraState {
+        case valid, notValid
+    }
+
+    lazy var previewView: CameraPreviewView = {
+        let previewView = CameraPreviewView()
+        previewView.translatesAutoresizingMaskIntoConstraints = false
+        (previewView.layer as? AVCaptureVideoPreviewLayer)?.videoGravity = .resize
+        let tapGesture = UITapGestureRecognizer(target: self, action: #selector(focusAndExposeTap))
+        previewView.addGestureRecognizer(tapGesture)
+        return previewView
+    }()
+    
+    init(giniConfiguration: GiniConfiguration = .shared) {
+        self.giniConfiguration = giniConfiguration
+        super.init(nibName: nil, bundle: nil)
+        
+        setupCamera(giniConfiguration: giniConfiguration)
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    deinit {
+        NotificationCenter.default.removeObserver(self)
```

Comment:
Is this really necessary? Wouldn't the notification center be smart enough to deal with it alone?

Reply:
Yes, unfortunately it has to be disposed manually.